src/mpid/common/thread/mpidu_thread_fallback.h

```diff
@@ -394,117 +306,61 @@ M*/
-            MPL_thread_id_t self_;                                      \
-            MPL_thread_self(&self_);                                    \
-            MPL_thread_same(&self_, &mutex.owner, &equal_);             \
-            if (equal_ && mutex.count > 0) {                            \
-                MPL_DBG_MSG(MPIR_DBG_THREAD, TYPICAL, "non-recursive yielding GLOBAL mutex"); \
-                MPL_DBG_MSG(MPIR_DBG_THREAD,VERBOSE,"enter MPIDU_Thread_yield"); \
-                MPIDU_Thread_yield(&mutex, &err_);                          \
-                MPL_DBG_MSG(MPIR_DBG_THREAD,VERBOSE,"exit MPIDU_Thread_yield"); \
-                MPIR_Assert(err_ == 0);                                 \
-            }                                                           \
-        }                                                               \
-    } while (0)
-#define MPIDUI_THREAD_CS_YIELD_VCI(mutex) MPIDUI_THREAD_CS_YIELD_GLOBAL(mutex)
-#define MPIDUI_THREAD_CS_YIELD_VCI_GLOBAL(mutex) do {} while (0)
-#define MPIDUI_THREAD_CS_YIELD_POBJ(mutex) do {} while (0)
+/* POBJ is only enabled with GRANULARITY__POBJ */
+#if MPICH_THREAD_GRANULARITY == MPICH_THREAD_GRANULARITY__POBJ
+#define MPIDUI_THREAD_CS_ENTER_POBJ(mutex)  MPIDUI_THREAD_CS_ENTER_NREC(mutex)
+#define MPIDUI_THREAD_CS_EXIT_POBJ(mutex)   MPIDUI_THREAD_CS_EXIT_NREC(mutex)
+#define MPIDUI_THREAD_CS_YIELD_POBJ(mutex)  MPIDUI_THREAD_CS_YIELD_NREC(mutex)
+#else
+#define MPIDUI_THREAD_CS_ENTER_POBJ(mutex)      /* NOOP */
+#define MPIDUI_THREAD_CS_EXIT_POBJ(mutex)       /* NOOP */
+#define MPIDUI_THREAD_CS_YIELD_POBJ(mutex)      /* NOOP */
+#endif
 
-#endif /* MPICH_THREAD_GRANULARITY */
+/* VCI is only enabled with MPICH_THREAD_GRANULARITY__VCI */
+#if MPICH_THREAD_GRANULARITY == MPICH_THREAD_GRANULARITY__VCI
+#define MPIDUI_THREAD_CS_ENTER_VCI(mutex) MPIDUI_THREAD_CS_ENTER_REC(mutex)
+#define MPIDUI_THREAD_CS_EXIT_VCI(mutex) MPIDUI_THREAD_CS_EXIT_REC(mutex)
+#define MPIDUI_THREAD_CS_YIELD_VCI(mutex) MPIDUI_THREAD_CS_YIELD_REC(mutex)
+#else
+#define MPIDUI_THREAD_CS_ENTER_VCI(mutex)       /* NOOP */
+#define MPIDUI_THREAD_CS_EXIT_VCI(mutex)        /* NOOP */
+#define MPIDUI_THREAD_CS_YIELD_VCI(mutex)       /* NOOP */
+#endif
 
-#else /* !defined(MPICH_IS_THREADED) */
+/* TRYENTER is of limited experimental use, only define the macros here when needed */
+#if MPICH_THREAD_GRANULARITY == MPICH_THREAD_GRANULARITY__VCI
+#define MPIDUI_THREAD_CS_TRYENTER_VCI(mutex,cs_acq) MPIDUI_THREAD_CS_TRYENTER_REC(mutex,cs_acq)
+#else
+#define MPIDUI_THREAD_CS_TRYENTER_VCI(mutex,cs_acq)     /* NOOP */
+#endif
 
-#define MPIDUI_THREAD_CS_YIELD_GLOBAL(mutex) do {} while (0)
-#define MPIDUI_THREAD_CS_YIELD_POBJ(mutex) do {} while (0)
+/* Stateful CS_ENTER/EXIT are only used in ch3:sock and only with GLOBAL */
+#define MPIDUI_THREAD_CS_ENTER_ST_GLOBAL    MPIDUI_THREAD_CS_ENTER_ST
+#define MPIDUI_THREAD_CS_EXIT_ST_GLOBAL     MPIDUI_THREAD_CS_EXIT_ST
```

Comment:
If we build `ch3` with the `POBJ` critical section, `CS_ENTER_ST_GLOBAL` would still point to `MPIDUI_THREAD_CS_ENTER_ST`. These two would need to go inside a check for thread granularity. Also, I would prefer it if we keep it in the same place as above where we define the rest of the macros depending on granularity. For the same reason as above: cleaner and easier to read.

Reply:
done